bftengine/tests/bcstatetransfer/source_selector_test.cpp

```diff
@@ -0,0 +1,241 @@
+
+constexpr uint64_t kSampleCurrentTimeMs = 1000;
+constexpr auto kSmallDeltaTime = kSampleCurrentTimeMs + 1;
+constexpr auto kBigDeltaTime = 10 * kSampleCurrentTimeMs;
+constexpr uint16_t kSampleReplicaId = 1;
+
+const auto replicas = std::set<uint16_t>{1, 2, 3, 4};
+
+TEST(source_selector_test, on_construction_there_is_no_current_replica) {
+  ASSERT_EQ(SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs).currentReplica(),
+            NO_REPLICA);
+}
+
+TEST(source_selector_test, has_no_source_on_construction) {
+  ASSERT_FALSE(SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs).hasSource());
+}
+
+TEST(source_selector_test, has_source_after_selection) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  source_selector.updateSource(kSampleCurrentTimeMs);
+  ASSERT_TRUE(source_selector.hasSource());
+}
+
+TEST(source_selector_test, zero_preferred_replica_on_construction) {
+  ASSERT_EQ(
+      SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs).numberOfPreferredReplicas(), 0);
+}
+
+TEST(source_selector_test, no_replica_is_preferred_on_construction) {
+  const auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  for (const auto& replica : replicas) {
+    ASSERT_FALSE(source_selector.isPreferred(replica));
+  }
+}
+
+TEST(source_selector_test, should_change_source_after_construction) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  // Make sure that the current replica is "NO_REPLICA".
+  // When there is no current replica the selector should select an initial replica as source.
+  ASSERT_EQ(source_selector.currentReplica(), NO_REPLICA);
+  ASSERT_TRUE(source_selector.shouldReplaceSource(kSampleCurrentTimeMs, false));
+}
+
+TEST(source_selector_test, should_change_source_when_bad_data_is_received_from_the_current_source) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  // Select a source
+  source_selector.updateSource(kSampleCurrentTimeMs);
+  // Ensuring that the change of the selected source is not caused by a timeout is
+  // done by intentionally passing a value than is not much bigger than the one stated as current time during the
+  // selection process.
+  ASSERT_TRUE(source_selector.shouldReplaceSource(kSmallDeltaTime, true));
+}
+
+TEST(source_selector_test, should_change_source_when_the_replacement_time_has_elapsed) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  source_selector.updateSource(kSampleCurrentTimeMs);
+  ASSERT_TRUE(source_selector.shouldReplaceSource(kBigDeltaTime, false));
+}
+
+TEST(source_selector_test, source_should_NOT_be_changed_when_working_and_replacement_time_has_NOT_elapsed_yet) {
```

Comment:
Nit: Maybe we can just use lower case for `NOT`. Not exactly sure what the coding standard for test names is, though. My personal preference is lower case.

Reply:
Just came to mind that we can shorten it by saying something like `source_should_not_be_changed_if_replacement_time_has_not_elapsed`.

Reply:
The casing has been taken into consideration and applied, however I think the name should still suggest that the source has to be working when looking for a timeout, otherwise it has to be changed regardless of the replacement time.